allennlp/data/dataset_readers/nlvr.py

```diff
@@ -0,0 +1,142 @@
+                 terminal_indexers: Dict[str, TokenIndexer] = None) -> None:
+        self._tokenizer = tokenizer or WordTokenizer()
+        self._sentence_token_indexers = sentence_token_indexers or {"tokens": SingleIdTokenIndexer()}
+        self._nonterminal_indexers = nonterminal_indexers or {"tokens":
+                                                              SingleIdTokenIndexer("rule_labels")}
+        self._terminal_indexers = terminal_indexers or {"tokens": SingleIdTokenIndexer("rule_labels")}
+        self._agenda_mapping: Dict[str, str] = {}
+        for constant in nlvr_types.COMMON_NAME_MAPPING:
+            alias = nlvr_types.COMMON_NAME_MAPPING[constant]
+            if alias in nlvr_types.COMMON_TYPE_SIGNATURE:
+                constant_type = nlvr_types.COMMON_TYPE_SIGNATURE[alias]
+                if constant_type != types.ANY_TYPE:
+                    self._agenda_mapping[constant] = "%s -> %s" % (constant_type, constant)
+
+    @overrides
+    def read(self, file_path):
+        instances = []
+        with open(file_path, "r") as data_file:
+            logger.info("Reading instances from lines in file: %s", file_path)
+            for line in tqdm.tqdm(data_file):
+                line = line.strip("\n")
+                if not line:
+                    continue
+                instances.append(self.text_to_instance(line))
+        if not instances:
+            raise ConfigurationError("No instances read!")
+        return Dataset(instances)
+
+    @overrides
+    def text_to_instance(self, json_line: str) -> Instance:
+        # pylint: disable=arguments-differ
+        data = json.loads(json_line)
+        sentence = data["sentence"]
+        label = data["label"] if "label" in data else None
+        structured_rep = data["structured_rep"]
+        world = NlvrWorld(structured_rep)
+        tokenized_sentence = self._tokenizer.tokenize(sentence)
+        sentence_field = TextField(tokenized_sentence, self._sentence_token_indexers)
+        agenda = self._get_agenda_for_sentence(sentence)
+        assert agenda, "No agenda found for sentence: %s" % sentence
+        is_nonterminal = lambda x: "<" not in x and len(x) > 1
+        production_rule_fields: List[Field] = []
+        instance_action_ids: Dict[str, int] = {}
+        for production_rule in world.all_possible_actions():
+            instance_action_ids[production_rule] = len(instance_action_ids)
+            field = ProductionRuleField(production_rule,
+                                        terminal_indexers=self._terminal_indexers,
+                                        nonterminal_indexers=self._nonterminal_indexers,
+                                        is_nonterminal=is_nonterminal,
+                                        context=tokenized_sentence)
+            production_rule_fields.append(field)
+        # agenda_field contains indices into actions.
+        agenda_field = ListField([LabelField(instance_action_ids[action],
```

Comment:
From what I understand from the code, I don't see a difference in the way `LabelField` is handling padding, from what I want. Can you tell what the difference is, or give a pointer to something I'm missing?

Reply:
`LabelField` assumes that you've got a string, and depending on values you pass in to the constructor, it might insert padding and UNK tokens into your ids.  `IndexField` is designed for exactly the use case you want here, `LabelField` wasn't.  Using `LabelField` is more error-prone.